core/trino-main/src/main/java/io/trino/client/direct/DirectTrinoClient.java

```diff
@@ -103,7 +113,27 @@ public DispatchQuery execute(SessionContext sessionContext, @Language("SQL") Str
                     Page page = pageDeserializer.deserialize(serializedPage);
                     queryResultsListener.consumeOutputPage(page);
                 }
-                getQueryFuture(whenAnyComplete(ImmutableList.of(queryManager.getStateChange(queryId, state), exchangeClient.isBlocked())));
+
+                ListenableFuture<Object> anyCompleteFuture = whenAnyComplete(ImmutableList.of(
+                        queryManager.getStateChange(queryId, state),
+                        exchangeClient.isBlocked()));
+                for (boolean anyComplete = false; !anyComplete; dispatchQuery.recordHeartbeat()) {
```

Comment:
nit: I think waiting for the `future.isDone()` instead of a separate variable would make the code a bit simpler
```
                while (!future.isDone()){
                    try {
                        future.get(heartBeatIntervalMillis, TimeUnit.MILLISECONDS);
                    }
                    catch (TimeoutException e) {
                        // continue waiting until the query state changes or the exchange client is blocked.
                        // we need to periodically record the heartbeat to prevent the query from being purged
                        dispatchQuery.recordHeartbeat();
                    }
...
```

Reply:
agreed. good call